Fix part of #130: Cellular data alert in audio player

rt4914 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

This PR contains the code for cellular data alert. Currently the code does not save user preference to storage.

## Output (low-fi)
<img width="341" alt="Screenshot 2019-09-19 at 10 19 01 PM" src="https://user-images.githubusercontent.com/9396084/65264100-83091d00-db2b-11e9-95e5-4c401c818ec2.png">


## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

Commit:
Celluar data alert in audio player

rt4914 commented:
@BenHenning @jamesxu0 PTAL

Commit:
Javadoc updates

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt:
```diff
@@ -8,21 +8,31 @@ import android.view.ViewGroup
 
-  lateinit var languageInterface: LanguageInterface
+  init {
+    // Add code to control the value of showCellularDataDialog from [PersistentCacheStore].
```
We should create a controller for this, or control this via a controller. PersistentCacheStore shouldn't be used directly by the UI.

@jamesxu0 FYI that this is another requirement you should consider when finishing the domain controllers for the audio player.

rt4914 replied:
I have updated the comment to `// Add code to control the value of showCellularDataDialog using AudioController.`

@jamesxu0 will be finishing this up in domain module.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt:
```diff
@@ -52,4 +62,25 @@ class AudioFragment : InjectableFragment() {
+    val dialogFragment = CellularDataDialogFragment.newInstance(
+      cellularDataInterface
+    )
+    // Might need to replace show() with showNow().
```
Let's use showNow(). I suspect it may not work because you're calling showCellularDataDialogFragment in onCreateView, and if I recall correctly you can't perform a fragment transaction during another fragment transaction.

We ought to not show a dialog immediately upon fragment creation. In fact, shouldn't the dialog only show when you try to play audio and you're on cellular?

rt4914 replied:
Regarding first point: Yes you are correct on fragment transaction point, it does it give error of fragment transaction already running.

Actually CellularDataDialogFragment should be visible even before we show AudioDialogFragment and based on user preference the audiobar will be visible.

I have put this here for testing purposes otherwise it might happen that when @jamesxu0 tries this code he may run into this error.

For now I will change this to `showNow` and add comment that for testing purpose only you can change this to `show`.

BenHenning replied:
Thanks for clarifying. After looking at the mocks again, it seems like we should this based on clicking the audio button in the player. I think that means we should be showing the cellular dialog at that point, and before we show the audio dialog. This also helps prevent the need to try and show a dialog during the lifecycle method of another fragment.

If the audio button checks for cellular connectivity and opens the cellular dialog, then it decides whether to proceed with showing the audio bar based on the user's input.

Do you agree?

rt4914 replied:
I agree on this completely, for now, I will add one dummy button on `StateFragment`and on clicking on that it will show `CellularDataDialogFragment` from that if the response favours showing of `AudioFragment` then I will show that otherwise I will not show that.

Later on that dummy button in `StateFragment` will be replaced with `item` in `toolbar`.

How does this sound?

BenHenning replied:
Yep sounds good--thanks!

rt4914 replied:
welcome.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataDialogFragment.kt:
```diff
@@ -0,0 +1,55 @@
+import org.oppia.app.R
+
+/**
+ * DialogFragment that controls user preference for dialog box
```
Nit: end with period for complete sentence.

Also, maybe something like the following would be clearer:

```DialogFragment that indicates to the user they are on cellular when trying to play an audio voiceover.```

rt4914 replied:
Updated comments to make it more descriptive.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataDialogFragment.kt:
```diff
@@ -0,0 +1,55 @@
+ * DialogFragment that controls user preference for dialog box
+ */
+class CellularDataDialogFragment : DialogFragment() {
+  private var doNotShowAlert: Boolean = false
```
Should this be ``doNotShowAgain``?

rt4914 replied:
Changed this variable to `doNotShowAgain` and that actually makes more sense now.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataInterface.kt:
```diff
@@ -0,0 +1,6 @@
+
+/** Interface to check the preference regarding alert for [CellularDataDialogFragment] */
+interface CellularDataInterface {
+  fun cellularDataUserPreference(doNotShowAlert: Boolean, playAudio: Boolean)
```
For perhaps a simpler interface, consider doubling the method based on the play boolean state:

```kotlin
enableAudioWhileOnCellular(doNotShowAgain: Boolean)
disableAudioWhileOnCellular(doNotShowAgain: Boolean)
```

rt4914 replied:
Changed the interface as per suggestions.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataDialogFragment.kt:
```diff
@@ -0,0 +1,55 @@
+    val checkBox = view.findViewById<CheckBox>(R.id.cellular_data_dialog_checkbox)
+
+    checkBox.setOnCheckedChangeListener { buttonView, isChecked ->
+      doNotShowAlert = isChecked
```
Do we need to retain this state on configuration changes?

rt4914 replied:
Actually, even if we save this or not this is blocked on configuration change support for `showCellularDataDialogFragment()`.

After running the app and changing the orientation of app, the app crashes with following message:
`FragmentManager is already executing transactions`
on `fragmentTransaction` in `showCellularDataDialogFragment`, so before fixing this `doNotShowAlert` variable, I will have to first fix this dialog.

BenHenning replied:
As discussed via chat, this is likely because you're creating the fragment in onCreateView(), and that will be called again during config changes. We can't perform fragment transactions (e.g. via showNow()) within another transaction (e.g. onCreateView() during config change).

Moving management of showing the dialog to a button in StateFragment should address this issue since it won't be attempting to show the dialog in a lifecycle method.

rt4914 replied:
Thanks for summarising this here.

rt4914 replied:
This approach is correct and is working correctly as per above discussed implementation.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt:
```diff
@@ -52,4 +62,25 @@ class AudioFragment : InjectableFragment() {
   }
+
+  private fun showCellularDataDialogFragment() {
+    cellularDataInterface = object : CellularDataInterface {
```
Shouldn't the fragment be implementing this interface? That seems cleaner to me.

rt4914 replied:
Unable to understand this point.

Do you mean that `Cellular DataDial Fragment` should implement this interface, if that is the case than I am confused how will be transfer data back to the AudioFragment once user has selected his preference.

Or

Do you mean something else?

BenHenning replied:
Sorry, what I mean is I think AudioFragment should implement this interface. That way the dialog fragment can do the equivalent:

```kotlin
((CellularDataInterface) getParentFragment()).enableAudioWhileOnCellular(...)
```

This is actually what I thought we were going to do when discussing this yesterday, per: https://stackoverflow.com/a/23144683/3689782.

rt4914 replied:
Okay, will change the implementation accordingly.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt:
```diff
@@ -52,4 +62,25 @@ class AudioFragment : InjectableFragment() {
+      cellularDataInterface
+    )
+    // Might need to replace show() with showNow().
+    dialogFragment.show(fragmentManager, TAG_CELLULAR_DATA_DIALOG)
```
We should also prevent this dialog from reshowing upon a configuration change since the user has already started playing audio, so the question is moot.

We may also need to retain this across states, but we should check with Chantel first on anticipated user behavior/expectations here. This second consideration shouldn't block this PR.

rt4914 replied:
Should I add this as a comment in this code and solve this once the discussion with chantel is done?

BenHenning replied:
It seems worthwhile to file the second part as an issue and add a TODO here (ie, whether we should ask for streaming on each state or one per exploration).

We should solve the configuration change case now, I think.

rt4914 replied:
Sounds great. will change this accordingly.

BenHenning reviewed (changes requested):
Generally LGTM @rt4914--can you also add tests to verify that the dialogs at least appear? I know we can't fully flesh out the behavioral aspects of when they should show up, but ideally if they are always showing then these tests will fail once the real domain implementations are added, and then we'll need to update them accordingly.

Commit:
Changed CellularDataInterface with simpler methods

Commit:
CellularDataFragment shifted to StateFragment

Commit:
Imrpovised Language and Celluar Fragments

Commit:
Implementation change in CellularDataINterface

Commit:
Implementation change in LanguageInterface

Commit:
Complete cellular, audio, language screen flow test

rt4914 commented:
> Generally LGTM @rt4914--can you also add tests to verify that the dialogs at least appear? I know we can't fully flesh out the behavioral aspects of when they should show up, but ideally if they are always showing then these tests will fail once the real domain implementations are added, and then we'll need to update them accordingly.

Test case done. This test case checks the entire flow of the audio component from showing `CelluarDataDialogFragment` to `AudioFragment` to `LanguageDialogFragment`

Commit:
Nit changes

rt4914 commented:
@BenHenning PTAL
I have made following changes as per our discussion:
1. Shifted code of `CellularDataDialogFragment` to `StateFragment`
2. Changed the usage implementation of `CellularDataInterface`and `LanguageInterface` as per your suggestions from this link:  https://stackoverflow.com/a/23144683/3689782
3. Added test-case which checks the entire flow of `CellularDataDialogFragment`, `AudioFragment` and `LanguageDialogFragment`
4. Based on user selection in `CellularDataDialogFragment`, the code controls whether to show `AudioFragment` or not.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt:
```diff
@@ -5,44 +5,58 @@ import android.os.Bundle
+    try {
+      languageInterface = fragment as LanguageInterface
+    } catch (e: ClassCastException) {
+      throw ClassCastException(
```
What's the purpose of this catch & rethrow?

rt4914 replied:
This was used from the reference provided in previous conversation https://stackoverflow.com/questions/23142956/sending-data-from-nested-fragments-to-parent-fragment/23144683#23144683

But after checking again, the code works even without using this.

rt4914 replied:
This `try-catch-throw` was mainly checking if the casting has been done correctly or not. And after checking the code correctly, we don't need this because our code is using LanguageInterface correctly.

Therefore removing this `try-catch-throw` is safe.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataDialogFragment.kt:
```diff
@@ -0,0 +1,59 @@
+      .setView(view)
+      .setMessage(R.string.cellular_data_alert_dialog_description)
+      .setPositiveButton(R.string.cellular_data_alert_dialog_okay_button) { dialog, whichButton ->
+        if(cellularDataInterface!=null) {
```
Nit: please reformat code to have proper spacing between elements per consistency with other classes in the codebase.

rt4914 replied:
Done here in everywhere else.

BenHenning reviewed (changes requested):
Just did a really quick glance, but didn't deep-dive yet. Got a bit tired all of a sudden, so I'll take a closer look at this later.

@jamesxu0 could you please give this PR a pass tomorrow?

Commit:
Nit changes and removed try-catch-throw code

jamesxu0 commented on app/src/main/java/org/oppia/app/player/state/StateFragment.kt:
```diff
@@ -5,19 +5,79 @@ import android.os.Bundle
+      fragmentManager?.beginTransaction()?.remove(previousFragment)?.commitNow()
+    }
+    val dialogFragment = CellularDataDialogFragment.newInstance(
+      cellularDataInterface
```
Is it necessary to make a cellularDataInterface variable since you can just pass in "this"?

rt4914 replied:
Correct. Updated this for cellular and language interface and removed unnecessary code.

Commit:
Removed unnecessary interface code

Commit:
Remove StateFragment usgae from HomeActivityController

Commit:
Nit javadoc changes

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataDialogFragment.kt:
```diff
@@ -0,0 +1,59 @@
+    val view = activity!!.layoutInflater.inflate(R.layout.cellular_data_dialog, null)
+    val checkBox = view.findViewById<CheckBox>(R.id.cellular_data_dialog_checkbox)
+
+    checkBox.setOnCheckedChangeListener { buttonView, isChecked ->
```
I'm actually wondering--do we need this listener? Could we just check the state of the checkBox upon closing the dialog to determine which value to pass along?

rt4914 replied:
Your point is correct. removed the `doNotShowAgain` variable and `setOnCheckedChangeListener ` and directly taking state of checkbox when needed.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataDialogFragment.kt:
```diff
@@ -0,0 +1,59 @@
+        }
+        dismiss()
+      }
+      .setCancelable(true)
```
What will cancellation do? Close the dialog without enabling/disabling audio?

rt4914 replied:
## Explanation

`setCancelable` is used so as to control the behaviour of dialog when user clicks on the default-android-back-button.
If it is `setCancelable(true)`, if user clicks on default-android-back-button when the dialog is shown, the dialog will dismiss itself.
If it is `setCancelable(false)`, if user clicks on default-android-back-button when the dialog is shown, nothing will happen and user is forced to select one of the options from the dialog.

## Code
Now, coming again to the code, I checked `.setCancelable(true)` and it is not working, so based on [this link](https://stackoverflow.com/questions/8906269/alertdialogs-setcancelablefalse-method-not-working), it suggests that I should set this true or false at the place where we are creating the dialogfragment.

I am removing this, because the default behaviour is `.setCancelable(true)`, so we won't need to set this property anywhere in the code and it will work just fine.

rt4914 replied:
This conversation can be resolved after @BenHenning looks at this conversation.

BenHenning replied:
Ah, I didn't know the nuance of setting it after creating, but my question was actually whether cancellation does the right thing here from the user's perspective. I know it automatically dismisses the dialog, but what will the user's experience be if they see this dialog and cancel it rather than dismiss/press okay?

rt4914 replied:
even if we use `setCancelable(true or false)` everything works fine because it does not work here in this case. That's why I have removed this. and the default behaviour of android on dialogs works correct for our case, which is, if the dialog is shown and user clicks on back button, the dialog will dismiss itself.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/CellularDataInterface.kt:
```diff
@@ -0,0 +1,7 @@
+
+/** Interface to check the preference regarding alert for [CellularDataDialogFragment]. */
+interface CellularDataInterface {
+  fun enableAudioWhileOnCellular(doNotShowAgain: Boolean)
```
In this context, maybe ``saveUserChoice`` would be a bit clearer on what needs to happen.

Also, it's a good practice to document the methods of an interface to explain when they should be called, and what implementations are expected to do in each situation (e.g. that if saveUserChoice is true, then the UI should go ahead and play audio in cellular cases and not re-prompt the user.

rt4914 replied:
Updated `doNotShowAgain ` to `saveUserChoice` here and elsewhere

Added comments for methods.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/LanguageDialogFragment.kt:
```diff
@@ -6,69 +6,47 @@ import android.os.Bundle
     /**
-     * This function is responsible for displaying content in DialogFragment
+     * This function is responsible for displaying content in DialogFragment.
      * @param languageInterface: [LanguageInterface] to send data back to parent
```
Nit: please add spacing between the ``@param`` line and the summary fragment above it.

rt4914 replied:
Added one extra line between description and `@param`

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/LanguageDialogFragment.kt:
```diff
@@ -6,69 +6,47 @@ import android.os.Bundle
+      languageList: List<String>,
       currentLanguageCode: String
     ): LanguageDialogFragment {
+      this.languageInterface = languageInterface
```
Instead of sharing static state, why not save the value directly in the fragment? Though, I think you using Bundle arguments makes a bit more sense since then the fragment will retain these values upon configuration changes.

rt4914 replied:
I tried that Bundle arguments code but actually, it is not able to transfer that data to the `onCreateView` inside the dialogFragment and because of which I shifted to this. I will try that once again, and post the exact error.

rt4914 replied:
I have updated the to work with Bundle and it is working now. Apparently, the accepted answers on stackoverflow were incorrect.

rt4914 replied:
Still not able to understand what you mean by this link
**Instead of sharing static state, why not save the value directly in the fragment?**

rt4914 replied:
I tried using `parentFragment` information for using languageInterface directly inside `LanguageInterfaceDialogFragment` but it is not working.

In my `LanguageInterfaceDialogFragment` the `parentFragment` is a `StateFragment`, but I think it should be `AudioFragment` because `AudioFragment` is using the `LanguageInterface`.

Similar, in my `CellularDataDialogFragment` the `parentFragment` is null, but here it should be `StateFragment` because `StateFragment` is using this. 

Do you know the reason behind this?

rt4914 replied:
I think for dialogFragments parentFragment information does not work.

Finally, I tried using `targetFragment` and it worked.

BenHenning replied:
My point on static state was around using a companion object since that's equivalent to using static variables to pass state in Java. Using a Bundle mitigates this concern.

Aha, I think it's because you're using supportFragmentManager and not childSupportFragmentManager. We should probably use child fragment manager since that treats the current fragment as a parent rather than the fragment's parent.

rt4914 replied:
Thanks for pointing out this mistake. `childSupportFragmentManager ` works.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragment.kt:
```diff
@@ -5,12 +5,25 @@ import android.os.Bundle
+  var isAudioFragmentShowing = ObservableField<Boolean>(false)
+
+  init {
+    // Add code to control the value of showCellularDataDialog using AudioController.
```
Please frame this as a TODO pointing to the correct GitHub issue to make this easier to track.

rt4914 replied:
Connected to issue #116

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragment.kt:
```diff
@@ -5,12 +5,25 @@ import android.os.Bundle
+  lateinit var stateFragmentPresenter: StateFragmentPresenter
+  // Control this boolean value from controllers in domain module.
+  private var showCellularDataDialog = true
+  var isAudioFragmentShowing = ObservableField<Boolean>(false)
```
I'm a bit not keen on exposing members of the fragment beyond callbacks. Maybe move this to a view model, instead, to keep the data structure corresponding to the current fragment separate from the fragment orchestration code itself.

rt4914 replied:
I have shifted this to `StateViewModel` but now the method calls have changed as described below:

Inside `StateFragment` `dummyButtonClicked()` -> `StatePresenter` `setAudioFragmentVisible()` -> `StateViewModel` `setAudioFragmentVisible()` and this final function updates the variable `isAudioFragmentShowing` which finally controls the UI.

BenHenning replied:
Although this was a bit of extra boilerplate, I find your current solution much cleaner. Thanks!

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/audio/CellularDataDialogFragmentTest.kt:
```diff
@@ -0,0 +1,72 @@
+  @Test
+  fun testCellularDataDialogFragment_loadDialogFragment_loadAudioFragment_loadLanguageFragment_isDisplayed() {
+    ActivityScenario.launch(HomeActivity::class.java).use {
+      Espresso.onView(withId(R.id.dummy_audio_button)).perform(click())
```
Nit: prefer statically importing onView() for consistency with other static imports, and to reduce the code that needs to be read/written here.

rt4914 replied:
Please check this updated code, if this is correct?

BenHenning replied:
LGTM.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/audio/CellularDataDialogFragmentTest.kt:
```diff
@@ -0,0 +1,72 @@
+
+/** Tests for [CellularDataDialogFragment]. */
+@RunWith(AndroidJUnit4::class)
+class CellularDataDialogFragmentTest {
```
Suggest also adding a TODO against #116 to properly test the flows where the user enables/disables on cellular with/without saving the setting (& for verifying that saving the setting works correctly).

rt4914 replied:
Added:
`// TODO(#116): Write correct test-cases when the user enables/disables on cellular with/without saving the setting.`

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragment.kt:
```diff
@@ -5,12 +5,25 @@ import android.os.Bundle
+  lateinit var stateFragmentPresenter: StateFragmentPresenter
+  // Control this boolean value from controllers in domain module.
+  private var showCellularDataDialog = true
+  var isAudioFragmentShowing = ObservableField<Boolean>(false)
```
Can you add tests to verify from the perspective of StateFragment that driving through the two dialog flows properly shows/hides the audio fragment in the expected situations?

rt4914 replied:
My test case inside [CellularDataDilaogFragmentTest] actually follows this flow:

1. Click on `dummyButton`
2. Click on `Ok` button in `CellularDataDialogFragment`
3. Check if `audio_fragment` is displayed or not
4. Click on `audio_language_text_view` inside AudioFragment
5. Check if `Audio Language` text is displayed().

Therefore this test case makes sure that both the dialogs are shown on proper clicks.

BenHenning replied:
I think my original comment wasn't very clear. I was actually looking for the two cellular dialog cases where pressing okay/cancel plays/doesn't play audio, but I realize we probably can't test that yet.

However, this test doesn't seem to test that the cellular dialog actually opens. Can you add tests to verify that the cellular dialog can open, and that pressing okay/cancel (2 tests) properly dismiss the dialog?

rt4914 replied:
Added test cases for two scenarios:

1. CellularDialogShows -> Click Ok
2. CellularDialogShows -> Click Cancel

BenHenning reviewed (changes requested):
Thanks @rt4914! Did a more thorough pass on this. PTAL at comments.

Commit:
Optimise DialogFragment code and test case

Commit:
Code optimization

Commit:
Updated isAudioFragmemntVisible usage

Commit:
Data share via Bundle in DialogFragment

Commit:
Merge remote-tracking branch 'upstream/develop' into exploration-player-audio-low-fi

Commit:
Using targetFragment for interfaces

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt:
```diff
@@ -27,22 +27,20 @@ class AudioFragment : InjectableFragment() {
+      selectedLanguageCode
     )
-    dialogFragment.showNow(fragmentManager, TAG_DIALOG)
+    dialogFragment.setTargetFragment(this, REQUEST_CODE_DIALOG)
```
This is cool, I didn't know about this. However, are we actually using the activity result? It doesn't seem like we need to set this here, but maybe I'm misunderstanding the intent.

Edit: ah I see you're setting this so that we can cast to the interface. If that's the only purpose of setting the target fragment, could we maybe instead use [getParentFragment()](https://developer.android.com/reference/android/support/v4/app/Fragment.html#getParentFragment())? That seems a bit cleaner since it avoids the extra onActivityResult behavior of setTargetFragment that we don't really need here.

rt4914 replied:
Actually, `REQUEST_CODE_DIALOG` inside `setTargetFragment` is not needed for us but `setTargetFragment` requires this parameter.

Regarding, `getParentFragment()` gives the reference to `StateFragment` rather than `AudioFragment` because of which I am not able to use `getParentFragment()`

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/LanguageDialogFragment.kt:
```diff
@@ -6,75 +6,55 @@ import android.os.Bundle
-      languageList.add("hi-en")
-    }
+    var selectedIndex = arguments!!.getInt(KEY_SELECTED_INDEX, 0)
+    val languageArrayList: ArrayList<String> = arguments?.getStringArrayList(KEY_LANGUAGE_LIST) as ArrayList<String>
```
Why is this cast needed? It seems superfluous since getStringArrayList() returns ArrayList<String>.

Also, the ``?.`` operator needed here? I think you can avoid it by saving arguments locally:

```kotlin
val args = checkNotNull(arguments) { "Expected arguments to be pass to LanguageDialogFragment" }
// args can now be used without the nullability considerations
```

rt4914 replied:
Regarding extra cast:

`val languageArrayList: ArrayList<String> = args.getStringArrayList(KEY_LANGUAGE_LIST) as ArrayList<String>`

If we do not add this extra cast then there is type mismatch warning
<img width="509" alt="Screenshot 2019-09-25 at 11 18 20 AM" src="https://user-images.githubusercontent.com/9396084/65572607-376cce00-df86-11e9-9f2f-8c9516e33ba5.png">
 

Added this line:

`val args = checkNotNull(arguments) { "Expected arguments to be pass to LanguageDialogFragment" }`

BenHenning reviewed (changes requested):
Thanks @rt4914! Generally LGTM, but I think there are still some earlier unresolved comments. Can you please take a look at those & address them? I'm specifically interested in the request for additional tests.

Commit:
Using childFragmentManager

BenHenning reviewed (approved):
Given that the remaining changes are trivial, feel free to submit once they're resolved (changing the fragment manager & adding the remaining 2 test cases).

rt4914 commented:
> Given that the remaining changes are trivial, feel free to submit once they're resolved (changing the fragment manager & adding the remaining 2 test cases).

Thanks @BenHenning for regular and prompt reviews.

I have replaced `fragmentManager` with `childFragmentManager` to establish parent/child relation within fragments.

Also, working on test-cases now.

Commit:
Test case for CellularDataDialogFragment

Commit:
Merge remote-tracking branch 'upstream/develop' into exploration-player-audio-low-fi

rt4914 commented:
Final task of adding test-cases has been done, therefore, merging this PR now.
